### PR TITLE
revert 4f4fd10 preserving fix for #260, fixing #319, #322 and #334

### PR DESCRIPTION
**Pull request by Connum**
This reverts the change from 4f4fd10 that caused at least issues #319, #322 and #334 (maybe even more...).
With a tiny change to the original code, it was confirmed in #260 that it still fixes that issue.

I also wrote a test for #319 locally, but wasn't sure whether we could use the sample PDF or not:
```php
/**
 * Test that issue related pdf can now be parsed
 *
 * @see https://github.com/smalot/pdfparser/issues/319
 */
public function testIssue319()
{
    $filename = $this->rootDir.'/samples/bugs/Issue319.pdf';

    $document = $this->fixture->parseFile($filename);

    $this->assertStringContainsString('HS technology s.r.o., IČO: 52081061, Sídlo: Fraňa Mojtu 22, 94901 Nitra,\n
    Slovenská republika\n
    \n
    \n
    99,10', $document->getText());
}
```

### CODE REVIEW

**Review by k00ni — CHANGES_REQUESTED** [review]
Thank you for your contribution! Your patch seems reasonable.

> I also wrote a test for #319 locally, but wasn't sure whether we could use the sample PDF or not:
* [ ] ~~What does that mean? Is it available now?~~

* [ ] ~~I saw that you asked around in the issues you referenced. Can you give me a short summary about what you found out? That would be helpful.~~

* [ ] ~~Please address my in-code comments.~~

**EDIT:** Solved in next review.

File: tests/Integration/ParserTest.php [hunk 1]
```diff
@@ -91,4 +91,34 @@ public function testIssue267()
...
+    /**
+     * Test that issue related pdf can now be parsed
+     * License of the content taken from https://stackoverflow.com in the sample PDF:
+     * CC BY-SA 2.5 https://creativecommons.org/licenses/by-sa/2.5/
+     *
+     * @see https://github.com/smalot/pdfparser/issues/334
+     */
+    public function testIssue334()
+    {
+        $filename = $this->rootDir.'/samples/bugs/Issue334.pdf';
```
Comment (k00ni) [inline tests/Integration/ParserTest.php:118]: That file contains screenshots from questions and answers out of Stackoverflow. I am not sure if you can include them in that state. Is there a way to make the PDF as small as possible and/or remove usernames and other copyright-related entities (user images). I am not an attorney, but I am afraid that might infringe copyright law.

**Comment by k00ni** [comment]
@romanmatyus (#319), @Karmel0x (#322) and @CharmaineLohSiYing (#334): It would be great if you could give us a short feedback, if this PR fixes your problems. Thank you in advance!

File: tests/Integration/ParserTest.php [hunk 2]
```diff
@@ -91,4 +91,34 @@ public function testIssue267()
         $this->assertEquals(Image::class, \get_class($document->getObjectById('128_0')));
         $this->assertStringContainsString('4 von 4', $document->getText());
     }
+
+    /**
+     * Test that issue related pdf can now be parsed
```
[the remark this answers is not in the document]
Comment (Connum) [reply tests/Integration/ParserTest.php:96]: Sure, I just copied that from another issue-specific test. I'll add a short description!

File: tests/Integration/ParserTest.php (hunk 1 above)
Comment (Connum) [reply tests/Integration/ParserTest.php:118]: You're probably right! The CC license only covers the content of the questions and answers. I'll try to reduce the sample PDF to the text.

**Comment by Connum** [comment]
> > I also wrote a test for #319 locally, but wasn't sure whether we could use the sample PDF or not:
> 
>     * [X]  What does that mean? Is it available now?
I can't include that test without a sample file to test against. It means I was able to verify that my changes fix the parsing of the file provided in #319, but we can't simply use that file without consent. I can try to reduce the file, but I fear changing and re-saving it could result in the original issue not occuring in the first place...

>     * [X]  I saw that you asked around in the issues you referenced. Can you give me a short summary about what you found out? That would be helpful.

I basically just confirmed that the issues were caused by the fix in 4f4fd10. In #260, the issue related to that fix, the author of the fix confirmed that with my changes their (not publically available) PDF still worked fine, affirming that we can have all three new issues fixed at the same time without reintroducing the issue raised in #260.

>     * [X]  Please address my in-code comments.
Done, I will address those in an additional commit.

File: tests/Integration/ParserTest.php (hunk 1 above)
Comment (Connum) [reply tests/Integration/ParserTest.php:118]: Oh, by the way...
> remove usernames

that would explicitly go against the attribution part of the CC license... 😉 But I'll try to strip off any images without losing the parts containing the issue (fingers crossed)

**Comment by CharmaineLohSiYing** [comment]
> @romanmatyus (#319), @Karmel0x (#322) and @CharmaineLohSiYing (#334): It would be great if you could give us a short feedback if this PR fixes your problems. Thank you in advance!

Hello! This PR fixes my problems. All the characters in my pdf file are now output properly just like in the demo. Thank you!

**Comment by Connum** [comment]
@k00ni: The sample PDF from #319 is signed, so it cannot be edited directly. Once I print it to a new PDF or import and export it with Illustrator, it does seem to introduce some new issues, but not necessarily the one we need to reproduce...

File: tests/Integration/ParserTest.php (hunk 1 above)
Comment (k00ni) [reply tests/Integration/ParserTest.php:123]: I was not talking about the PDF of #319, but #334!

Comment (Connum) [reply tests/Integration/ParserTest.php:123]: Maybe I got you wrong, but well, we talked about both of them really, but inside this review conversation, it's all about #334. And the CC license applies to that PDF.

_(Sorry for resolving this, I'm not too familiar with all the processes of OS collaboration, and I thought that regarding #334, everything was now ok, as I reduced the sample via the additional commit)_

Comment (k00ni) [reply tests/Integration/ParserTest.php:123]: Based on the following text from Stackoverflow (Soure: https://stackoverflow.com/help/licensing):

```
As noted in the Stack Exchange Terms of Service and in the footer of every page, all publicly 
accessible user contributions are licensed under Creative Commons Attribution-ShareAlike 
license as follows:

* Content contributed before 2011-04-08 (UTC) is distributed under the terms of CC BY-SA 2.5.

* Content contributed from 2011-04-08 up to but not including 2018-05-02 (UTC) is distributed under 
the terms of CC BY-SA 3.0.

* Content contributed on or after 2018-05-02 (UTC) is distributed under the terms of CC BY-SA 4.0.
The license applicable for each Question and Answer revision is available on the post timeline. 
See this post for more information.

Please read the terms of service and the full legal text of the license carefully for more details on 
how your content can be used and for how you can use publicly accessible content contributed to 
the site by other users.
```
The content of that file we talked about falls under the CC [BY-SA 2.5](https://creativecommons.org/licenses/by-sa/2.5/) (its older than 2011-04-08).

CC BY-SA allows us to "remix, transform, and build upon the material for any purpose, even commercially.", therefore removals of usernames are allowed. Anyway, I did some research and found that usernames are not "creative work", therefore not subject for copyright law.

So file is OK and subject solved. Thank you for helping out here.

**Review by k00ni — APPROVED** [review]
> The sample PDF from #319 is signed, so it cannot be edited directly. Once I print it to a new PDF or import and export it with Illustrator, it does seem to introduce some new issues, but not necessarily the one we need to reproduce...

We can't use it, because it contains a logo, but its optional for this PR so we can ignore it.

* [x] I saw that you asked around in the issues you referenced. Can you give me a short summary about what you found out? That would be helpful.

Solved :+1:

**Comment by k00ni** [comment]
I will keep that open for a ~week to allow others to comment. If there are no objections, I will merge.

**Comment by MrMadClown** [comment]
Hi,
might be related to [4f4fd10](https://github.com/smalot/pdfparser/commit/4f4fd108270c32c0e81ea3dceaac2249162f27dc).
The text retrieved from [This PDF](https://www.btg-bestellservice.de/pdf/10070000.pdf) is pretty scrambled in the current version (0.16.2). I reverted to 0.14.0 (pre 4f4fd10) no more scrambling.
I'm still seeing way more spaces in the parsed Text than I can see in the PDF but those might actually be zero-width-spaces not sure.

**Comment by k00ni** [comment]
Can you please check if you have to same behavior with the version in this branch (https://github.com/Connum/pdfparser/tree/fix-260-319-322-334)?

**Comment by Connum** [comment]
I can confirm that the parsing of the PDF provided by @MrMadClown is also fixed with this PR! One example:

Before
```
Gestaltung und Sat4GIGNKPFKU9GUVRJCN)TC�M&GUKIn / Norbert Lauterbach
```

After
```
Gestaltung und Satz: Regelindis Westphal Grafik-Design / Norbert Lauterbach
```

All other instances of scrambled text are fixed as well, no more "strange" characters!

**Comment by k00ni** [comment]
Thank you @Connum and the others for your effort.

**Pull request merged**